### ISSUE TITLE
poetry config not working for HTTP auth so unable to pull private package

### ISSUE THREAD

**Issue by viveksair** [python-poetry/poetry#2125]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [ ] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: 
``` 
debian:10 
root@91ca4e71776b:/# uname -a
Linux 91ca4e71776b 4.9.93-linuxkit-aufs #1 SMP Wed Jun 6 16:55:56 UTC 2018 x86_64 GNU/Linux
```
- **Poetry version**:
```
poetry --version
Poetry version 1.0.3
```
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: 
```
  
[tool.poetry]
name = "foo_bar"
version = "0.1.0"
description = ""
authors = ["Your Name <you@example.com>"]

[tool.poetry.dependencies]
python = "^3.7"
private_package = "^1.0.13"


[tool.poetry.dev-dependencies]



[tool.poetry.scripts]


[[tool.poetry.source]]
name = "foo"
url = "https://foo.bar/simple"
default = true		

[build-system]
requires = ["poetry>=1.0.3"]
build-backend = "poetry.masonry.api"
```

## Issue

```
$ poetry config repositories. foo https://foo.bar/simple
$ poetry config $USER_ID $USER_PASSWORD
$ poetry add private_package

[ValueError]
Could not find a matching version of package private_package

```
and I tried with env as well ```POETRY_HTTP_BASIC_FOO_USERNAME POETRY_HTTP_BASIC_FOO_PASSWORD``` but not working.

but `poetry add private_package` working after mounting config & password at ``` ~/.config/pypoetry/auth.toml & ~/.config/pypoetry/config.toml ```

problem with mounting config is throwing user and password when error occurred while `poetry install` and I have to add `default = true` to `[[tool.poetry.source]]` to my  pyproject.toml

**Comment by 4F2E4A2E** [comment]
Thank you poetry for being awesome and thanks @viveksair for creating this issue.

I also can confirm this bug.

**Python version: 3.6.8
Poetry Version: 1.0.5**
**Install state**: fresh install, everything from scratch
**Auth**: basic-auth (via env like @viveksair did, as config command, or in the url as param `https://admin@admin123:foo.bar/simple`) did not made any difference. No special chars in the password were used, just: `admin123`
**Repo**: I am using nexus as a repo, poetry publish is working fine via token auth, but installing the private_package is not working in any way. There are no typos and passing the exact version of the package none or the wildcard ('*') did not help.

**pyproject.toml**
```toml
[[tool.poetry.source]]
name = "foo"
url = "https://foo.bar/simple"
secondary = true
```

**poetry add**
```bash
poetry add -vvv private_package
Using virtualenv: /Users/foo/.pyenv/versions/3.6.8/envs/py368_tf111
PyPI: No packages found for private_package *
anofau-test: 0 packages found for private_package *

[ValueError]
Could not find a matching version of package private_package

Traceback (most recent call last):
  File "/Users/foo/.poetry/lib/poetry/_vendor/py3.6/clikit/console_application.py", line 131, in run
    status_code = command.handle(parsed_args, io)
  File "/Users/foo/.poetry/lib/poetry/_vendor/py3.6/clikit/api/command/command.py", line 120, in handle
    status_code = self._do_handle(args, io)
  File "/Users/foo/.poetry/lib/poetry/_vendor/py3.6/clikit/api/command/command.py", line 171, in _do_handle
    return getattr(handler, handler_method)(args, io, self)
  File "/Users/foo/.poetry/lib/poetry/_vendor/py3.6/cleo/commands/command.py", line 92, in wrap_handle
    return self.handle()
  File "/Users/foo/.poetry/lib/poetry/console/commands/add.py", line 89, in handle
    packages, allow_prereleases=self.option('allow-prereleases')
  File "/Users/foo/.poetry/lib/poetry/console/commands/init.py", line 302, in _determine_requirements
    requirement['name'], allow_prereleases=allow_prereleases
  File "/Users/foo/.poetry/lib/poetry/console/commands/init.py", line 338, in _find_best_version_for_package
    'Could not find a matching version of package {}'.format(name)
```
https://github.com/python-poetry/poetry/blob/12db4a5e189ea8514e70f22279ae540274a0e5af/poetry/console/commands/init.py#L338

**Comment by saisasidhar** [comment]
I ran into this exact problem(same error message) several times while trying out poetry. I was surprised that the private repositories feature had been implemented for quite some time and release versions have all included tests for this use-cases, yet it did not work as expected. 

Atleast in my case, I ran into this problem because I was using `devpi` in a peculiar way.

The actual simple index (PEP503) html page of devpi server is at [http://pypi.example.com/project/dev/+simple/](http://pypi.example.com/project/dev/+simple/). Note the `+simple` in path.

If one were to use full path to this html page in `pyproject.toml`, poetry would not have any problem searching, locking and installing private packages.

This is the way it is because, `poetry.repositories.legacy_repository.Page` searches the html page for anchor tags **that have only a specific set of supported extensions** (.tar.gz, .whl etc).

My assumption that using [http://pypi.example.com/project/dev/](http://pypi.example.com/project/dev/) would just work stems from pip. 😄 

I looked at the verbose logs of `pip install .. --extra-index-url=...`, and noticed that pip does package search a bit differently.  It directly attempts to GET the url: `http://pypi.example.com/project/dev/package` which returns a 302 redirect to `http://pypi.example.com/project/dev/+simple/package/`. This doesn't seem to be the case when I try curl/postman. I'm not sure why !

Similarly, these url without `+simple` also worked as expected in pipenv, because pip is used in the backend

**Comment by hfjn** [comment]
Big, big thanks for poetry! ❤️ 

I have the exact same problem with nexus. Funnily enough pip is able to locate the image when using the exact same user, url and password:
```
pip search --index https://nexus.company/repository/pip/pypi package
User for nexus.company: user
Password:
package (0.1.0)  -
```

Poetry just spits out that I wasn't able to locate anything.

**Comment by bneijt** [comment]
When using a source with basic auth without configuring the credentials correctly, the source is silently seen as empty.

At https://github.com/python-poetry/poetry/blob/dcf30073d24c29b2d53932ba8599ff049f3f2bb8/poetry/repositories/legacy_repository.py#L365

there is only a check for 404. Meanwhile 401 is parsed, no content is found (it's a you have to log in page) and an empty list is returned, resulting in a lock simply saying there is a solver problem and no valid version is found.

After some trial and error, I was able to get poetry to pick up the basic auth configuration and see dependencies. The main issue here is that there should be a check for 200 and if not, throw an exception.

**Comment by trallnag** [comment]
I'm having the same issue... `poetry config http-basic.myrepo username password` is simply ignored. When I perform the install with `pip install --extra-index-url` it works fine. Bummer

**Comment by trallnag** [comment]
Difference is for me it works putting the credentials directly into the url in pyproject toml

**Comment by neersighted** [comment]
There was an issue with the environment variables not being respected that is long fixed on 1.2.x -- please open a new issue if you are setting the variables correctly and run into issues.
